src/compiler/checker.ts

```diff
@@ -11749,23 +11749,23 @@ module ts {
             return !!resolveName(location, name, SymbolFlags.Value, /*nodeNotFoundMessage*/ undefined, /*nameArg*/ undefined);
         }
 
-        function getBlockScopedVariableId(n: Identifier): number {
+        function getNamedValueId(n: Identifier, meaning: SymbolFlags): number {
             Debug.assert(!nodeIsSynthesized(n));
 
-            let isVariableDeclarationOrBindingElement =
-                n.parent.kind === SyntaxKind.BindingElement || (n.parent.kind === SyntaxKind.VariableDeclaration && (<VariableDeclaration>n.parent).name === n);
+            let isDeclarationOrBindingElement =
+                n.parent.kind === SyntaxKind.BindingElement || (isDeclarationName(n) && n.parent.kind !== SyntaxKind.ShorthandPropertyAssignment);
 
             let symbol = 
-                (isVariableDeclarationOrBindingElement ? getSymbolOfNode(n.parent) : undefined) ||
+                (isDeclarationOrBindingElement ? getSymbolOfNode(n.parent) : undefined) ||
                 getNodeLinks(n).resolvedSymbol ||
                 resolveName(n, n.text, SymbolFlags.Value | SymbolFlags.Alias, /*nodeNotFoundMessage*/ undefined, /*nameArg*/ undefined);
 
-            let isLetOrConst =
+            let hasMatchingMeaning =
                 symbol &&
-                (symbol.flags & SymbolFlags.BlockScopedVariable) &&
-                symbol.valueDeclaration.parent.kind !== SyntaxKind.CatchClause;
+                (symbol.flags & meaning) &&
+                (!(meaning & SymbolFlags.BlockScopedVariable) || symbol.valueDeclaration.parent.kind !== SyntaxKind.CatchClause);

```

Comment:
I don't think the check for `meaning & SymbolFlags.BlockScopedVariable` is necessary. The only way for the parent to be a catch clause is if it is also a block scoped variable.